### PR TITLE
Various fixes/workarounds to kdd based policy and namespace selection

### PR DESCRIPTION
**Pull request by robbrockbank**
Bug fixes:

-  Wasn't handling namespace selection correctly (was only in KDD code, but is now non-KDD only)
-  Policy name from k8s policy should not have the namespace in it anymore
-  Added orchestrator label so that k8s-backed policy only applies to k8s WEPs.

Also:
-  Some name prefix changes

### CODE REVIEW

File: lib/backend/k8s/conversion/constants.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conversion
+
+const (
+	NamespaceLabelPrefix       = "pcns."
+	Orchestrator               = "k8s"
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/constants.go]: For better readability I think this should be `OrchestratorKubernetes` or similar.

Probably fine for now since there's only one value, but we might eventually have `OrchestratorMesos` etc

Comment (robbrockbank) [reply lib/backend/k8s/conversion/constants.go]: Well if we keep the constant here then it is scoped by teh KDD package

If you want me to move to apiv2 then I'll Add a set of values and make this OrchestratorKubernetes

Comment (robbrockbank) [reply lib/backend/k8s/conversion/constants.go]: just say the word and i'll do either

Comment (robbrockbank) [reply lib/backend/k8s/conversion/constants.go]: Actually it should be in apiv2!  I'll move it!

File: lib/backend/k8s/conversion/conversion.go [hunk 2]
```diff
@@ -239,7 +226,7 @@ func (c Converter) PodToWorkloadEndpoint(pod *kapiv1.Pod) (*model.KVPair, error)
 // K8sNetworkPolicyToCalico converts a k8s NetworkPolicy to a model.KVPair.
 func (c Converter) K8sNetworkPolicyToCalico(np *extensions.NetworkPolicy) (*model.KVPair, error) {
 	// Pull out important fields.
-	policyName := fmt.Sprintf("knp.default.%s.%s", np.ObjectMeta.Namespace, np.ObjectMeta.Name)
+	policyName := fmt.Sprintf( K8sNetworkPolicyNamePrefix + np.ObjectMeta.Namespace)
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion.go]: looks like an extra space here

Comment (caseydavenport) [reply lib/backend/k8s/conversion/conversion.go]: Plus this should be `Name` not `Namespace`

File: lib/backend/k8s/conversion/conversion.go [hunk 3]
```diff
@@ -313,14 +300,16 @@ func (c Converter) K8sNetworkPolicyToCalico(np *extensions.NetworkPolicy) (*mode
...
-		selectors = append(selectors, fmt.Sprintf("calico/k8s_ns == '%s'", *ns))
+func (c Converter) k8sSelectorToCalico(s *metav1.LabelSelector, selectorType selectorType) string {
+	// If this is a namespace selector, the labels need to be prefixed with the profile namespace
+	// prefix.
+	var selectors []string
+	var prefix string
+	if selectorType == SelectorNamespace {
+		prefix = NamespaceLabelPrefix
+	}
+	if selectorType == SelectorPod {
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion.go]: This isn't quite right - the pod selector needs to be ANDed with that pod's namespace as well, not just with the entire cluster.

We still need this selector: `("projectcalico.org/namespace == '%s', *ns)`

Comment (caseydavenport) [reply lib/backend/k8s/conversion/conversion.go]: Also, I think _all_ selectors generated by this function need the `projectcalico.org/orchestrator == 'k8s'` bit.

Comment (robbrockbank) [reply lib/backend/k8s/conversion/conversion.go]: Ah but policy now has a namespace, so this rule does not need to be part of the policy selector.

However, instead - in the conversion to the v1 model - that's where we insert the additional selector to match on the namespace.

Comment (robbrockbank) [reply lib/backend/k8s/conversion/conversion.go]: (at least that's my thinking)

Comment (caseydavenport) [reply lib/backend/k8s/conversion/conversion.go]: Ok, so looks like the `namespace == foo` bit is added in the networkpolicy processor, is that right?

That makes sense.  I still think all of them need the orchestrator label

File: lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go [hunk 4]
```diff
@@ -58,11 +59,22 @@ func convertPolicyV2ToV1Spec(spec apiv2.PolicySpec) (interface{}, error) {
 		erules = append(erules, RuleAPIV2ToBackend(erule))
 	}
 
+	// If this policy is namespaced, then add a namespace selector.
+	selector := spec.Selector
+	if ns != "" {
+		nsSelector := fmt.Sprintf("%s == '$s'", apiv2.NamespaceLabel, ns)
```
Comment (caseydavenport) [inline lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go]: making a note: we're going to need corresponding changes to the CNI plugin to ensure it's writing the correct labels.

Comment (caseydavenport) [reply lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go]: This should be a `%s` not a `$s`

File: lib/backend/k8s/conversion/conversion.go (hunk 3 above)
Comment (robbrockbank) [reply lib/backend/k8s/conversion/conversion.go]: yup - i think you are correct about the orchestrator -- adding now

File: lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go [hunk 5]
```diff
@@ -114,6 +114,11 @@ func convertWorkloadEndpointV2ToV1Value(val interface{}) (interface{}, error) {
 		})
 	}
 
+	// Felix does not have the concept of namespaces, so add the namespace as a Label.  We'll
```
Comment (caseydavenport) [inline lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: Hm, I wasn't expecting this code here, but I'm not sure what to make of it.

My expectation was that each plugin would write the namespace in.  This is nice because then they don't have to, but it also makes for an asymmetrical UX between defining a policy using certain label and being unable to query and see that label in, say `calicoctl`.

File: lib/backend/k8s/conversion/conversion.go [hunk 6]
```diff
@@ -207,6 +190,10 @@ func (c Converter) PodToWorkloadEndpoint(pod *kapiv1.Pod) (*model.KVPair, error)
 		}
 	}
 
+	// Add the orchestrator label to the workload endpoint labels.
+	labels := pod.ObjectMeta.Labels
+	labels[apiv2.OrchestratorLabel] = Orchestrator
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion.go]: I think it would be good to show the Namespace label here as well - that way users can see that the wep has that label on it.

File: lib/backend/k8s/conversion/conversion.go [hunk 7]
```diff
@@ -207,6 +190,10 @@ func (c Converter) PodToWorkloadEndpoint(pod *kapiv1.Pod) (*model.KVPair, error)
 		}
 	}
 
+	// Add the orchestrator label to the workload endpoint labels.
+	labels := pod.ObjectMeta.Labels
+	labels[apiv2.OrchestratorLabel] = apiv2.OrchestratorKubernetes
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion.go]: I think your latest round erased my comment here, but I think it would be good to add the Namespace label to this map as well.

It means that people who query this through calicoctl will see it.  We probably want to do that in the CNI plugin as well so it works with etcd.

File: lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go (hunk 5 above)
Comment (caseydavenport) [reply lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: If anything, this logic should go in the defaulting for WorkloadEndpoints so that whenever a WEP is written it gets this label on it (plus PodToWorkloadEndpoint for KDD).

That's a better approach I think.

Comment (robbrockbank) [reply lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: hmm not sure i agree about not writing in the namespace.

Otherwise the concept of namespace vs global resources is completely dependent on the behavior of the plugin - that doesn't make sense to me.

Eventually we should add the concept of namespace to felix, so we can have namespace selectors and then all this label manipulation for the namespace goes away - that's not the case if the orchestrator is trying to control that.

Comment (robbrockbank) [reply lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: I could possible buy that we return the namespace label in the policy definition - but that would be nothing to do with the orchestrator though - more just something in the client to make it clear to users what is going on.

Comment (robbrockbank) [reply lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: In fact I would go one further and say that we should be adding the Orchestrator label here too, rather than within the KDD code - since we know the orchestrator from the name/Spec.

And in the client Get/List code insert the same labels so that people can at least see labels we are "auto-inserting".

If it isn't added generically then the behavior of what the namespace implies is different depending on the orchestrator - but I think we probably want that behavior defined.

File: lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go [hunk 8]
```diff
@@ -114,17 +114,26 @@ func convertWorkloadEndpointV2ToV1Value(val interface{}) (interface{}, error) {
...
+	if labels == nil {
+		labels = make(map[string]string, 2)
+	}
+	labels[apiv2.LabelNamespace] = v2res.Namespace
+	labels[apiv2.LabelOrchestrator] = v2res.Spec.Orchestrator
+
 	v1value := &model.WorkloadEndpoint{
 		State:            "active",
 		Name:             v2res.Spec.InterfaceName,
-		ActiveInstanceID: v2res.Spec.ContainerID,
```
Comment (caseydavenport) [inline lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: Why this change?

Comment (robbrockbank) [reply lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: It's not required for Felix.

Comment (caseydavenport) [reply lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: Yep. I was thinking along the lines of "what if another syncer client comes along that wants it" - e.g. I'd like to eventually update the kube-controllers code to use a syncer rather than a periodic poll, and it does care about this field.

I think I wouldn't use an updateprocessor to convert to v1 in that case though, so it's probably fine.

File: lib/backend/k8s/conversion/conversion.go [hunk 9]
```diff
@@ -171,13 +162,6 @@ func (c Converter) PodToWorkloadEndpoint(pod *kapiv1.Pod) (*model.KVPair, error)
 	// the host-side veth configured by the CNI plugin.
 	interfaceName := VethNameForWorkload(wepName)
 
-	// Build the labels map.
-	labels := map[string]string{}
-	if pod.ObjectMeta.Labels != nil {
-		labels = pod.ObjectMeta.Labels
-	}
-	labels["calico/k8s_ns"] = pod.ObjectMeta.Namespace
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion.go]: I think we still need to set the "magic" labels in this function - there are other places in our code - probably just kube-controllers - that relies on this function returning the same set of labels that you would see when doing a List/Get through libcalico.

File: lib/backend/k8s/conversion/conversion_test.go [hunk 10]
```diff
@@ -410,12 +391,12 @@ var _ = Describe("Test NetworkPolicy conversion", func() {
 		By("parsing the policy", func() {
 			pol, err = c.K8sNetworkPolicyToCalico(&np)
 			Expect(err).NotTo(HaveOccurred())
-			Expect(pol.Key.(model.ResourceKey).Name).To(Equal("knp.default.default.testPolicy"))
+			Expect(pol.Key.(model.ResourceKey).Name).To(Equal("knp.default.testPolicy"))
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion_test.go]: We should add in an assertion on the Namespace being "default" as well.

Comment (caseydavenport) [reply lib/backend/k8s/conversion/conversion_test.go]: And related, make sure we have a test that checks the updateprocessor correctly namespaces the selectors when given a NetworkPolicy (if we don't already)

File: lib/backend/k8s/conversion/conversion.go (hunk 9 above)
Comment (robbrockbank) [reply lib/backend/k8s/conversion/conversion.go]: sounds good.

in that case - i'll just make sure we set the labels in the clientv2 as well and persist them and then remove the syncer magic.

File: lib/backend/k8s/conversion/conversion_test.go (hunk 10 above)
Comment (robbrockbank) [reply lib/backend/k8s/conversion/conversion_test.go]: Good call!  Uncovered a conversion bug.

File: lib/backend/k8s/conversion/conversion_test.go [hunk 11]
```diff
@@ -1049,6 +1054,10 @@ var _ = Describe("Test NetworkPolicy conversion (k8s <= 1.7, no policyTypes)", f
 			pol, err = c.K8sNetworkPolicyToCalico(&np)
 			Expect(err).NotTo(HaveOccurred())
 			Expect(pol.Key.(model.ResourceKey).Name).To(Equal("knp.default.testPolicy"))
+			Expect(pol.Key.(model.ResourceKey).Namespace).To(Equal("default"))
+			Expect(pol.Value.(*apiv2.NetworkPolicy).Name).To(Equal("knp.default.testPolicy"))
+			Expect(pol.Value.(*apiv2.NetworkPolicy).Namespace).To(Equal("default"))
+			Expect(int(*pol.Value.(*apiv2.NetworkPolicy).Spec.Order)).To(Equal(1000))
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion_test.go]: Looks like a duplicate Expect here

Comment (robbrockbank) [reply lib/backend/k8s/conversion/conversion_test.go]: one is checking the key, the other the metadata

File: lib/backend/k8s/conversion/conversion.go [hunk 12]
```diff
@@ -218,10 +211,10 @@ func (c Converter) PodToWorkloadEndpoint(pod *kapiv1.Pod) (*model.KVPair, error)
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      wepName,
 				Namespace: pod.ObjectMeta.Namespace,
-				Labels:    labels,
+				Labels:    pod.ObjectMeta.Labels,
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion.go]: nit, but just `pod.Labels` seems to be what we do elsewhere.

File: lib/backend/k8s/conversion/conversion_test.go [hunk 13]
```diff
@@ -227,7 +213,7 @@ var _ = Describe("Test Pod conversion", func() {
 		Expect(wep.Value.(*apiv2.WorkloadEndpoint).Spec.Endpoint).To(Equal("eth0"))
 		Expect(wep.Value.(*apiv2.WorkloadEndpoint).Spec.Orchestrator).To(Equal("k8s"))
 		Expect(len(wep.Value.(*apiv2.WorkloadEndpoint).Spec.IPNetworks)).To(Equal(1))
-		Expect(wep.Value.(*apiv2.WorkloadEndpoint).ObjectMeta.Labels).To(Equal(map[string]string{"calico/k8s_ns": "default"}))
+		Expect(wep.Value.(*apiv2.WorkloadEndpoint).ObjectMeta.Labels).To(BeNil())
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion_test.go]: Don't think this test is right - should have the orchestrator / namespace labels yeah?

File: lib/backend/k8s/conversion/conversion_test.go (hunk 10 above)
Comment (caseydavenport) [reply lib/backend/k8s/conversion/conversion_test.go]: @robbrockbank Did you add this test? I don't see it.

Comment (caseydavenport) [reply lib/backend/k8s/conversion/conversion_test.go]: Oh it's way below. ok.

File: lib/backend/k8s/conversion/conversion.go (hunk 12 above)
Comment (robbrockbank) [reply lib/backend/k8s/conversion/conversion.go]: yup - was fixing that up where I noticed - guess this one slipped by

File: lib/backend/k8s/conversion/conversion_test.go [hunk 14]
```diff
@@ -1067,12 +1053,16 @@ var _ = Describe("Test NetworkPolicy conversion (k8s <= 1.7, no policyTypes)", f
 		By("parsing the policy", func() {
 			pol, err = c.K8sNetworkPolicyToCalico(&np)
 			Expect(err).NotTo(HaveOccurred())
-			Expect(pol.Key.(model.ResourceKey).Name).To(Equal("knp.default.default.testPolicy"))
+			Expect(pol.Key.(model.ResourceKey).Name).To(Equal("knp.default.testPolicy"))
+			Expect(pol.Key.(model.ResourceKey).Namespace).To(Equal("default"))
+			Expect(pol.Value.(*apiv2.NetworkPolicy).Name).To(Equal("knp.default.testPolicy"))
+			Expect(pol.Value.(*apiv2.NetworkPolicy).Namespace).To(Equal("default"))
+			Expect(int(*pol.Value.(*apiv2.NetworkPolicy).Spec.Order)).To(Equal(1000))
```
Comment (caseydavenport) [inline lib/backend/k8s/conversion/conversion_test.go]: Duplicate Expect()

File: lib/backend/k8s/conversion/conversion_test.go (hunk 13 above)
Comment (robbrockbank) [reply lib/backend/k8s/conversion/conversion_test.go]: well that is sad.  Let me investigate.

File: lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go [hunk 15]
```diff
@@ -58,11 +59,22 @@ func convertPolicyV2ToV1Spec(spec apiv2.PolicySpec) (interface{}, error) {
...
 	}
 
+	// If this policy is namespaced, then add a namespace selector.
+	selector := spec.Selector
+	if ns != "" {
+		nsSelector := fmt.Sprintf("%s == '%s'", apiv2.LabelNamespace, ns)
+		if selector == "" {
+			selector = nsSelector
+		} else {
+			selector += " && " + nsSelector
```
Comment (caseydavenport) [inline lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go]: I don't _think_ we generate any `||` clauses for these selectors.

If we did, we'd need `selector = fmt.Sprintf("%s && (%s)", nsSelector, selector)` or similar.

Comment (caseydavenport) [reply lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go]: To make sure the grouping is correct.

File: lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go [hunk 16]
```diff
@@ -114,10 +114,12 @@ func convertWorkloadEndpointV2ToV1Value(val interface{}) (interface{}, error) {
 		})
 	}
 
+	// Add the namespace and orchestrator labels to the set of labels we give Felix.  We
+	// don't need to persist these, and we need to make sure they are not user-configurable
```
Comment (caseydavenport) [inline lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: Think this comment is out of date now

Comment (robbrockbank) [reply lib/backend/syncersv1/updateprocessors/workloadendpointprocessor.go]: yup.  ugh tired.

File: lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go (hunk 15 above)
Comment (robbrockbank) [reply lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go]: yes i agree - and I think we DO need this for the user defined policy case

File: lib/backend/syncersv1/updateprocessors/rules.go [hunk 17]
```diff
@@ -46,6 +49,45 @@ func RuleAPIV2ToBackend(ar apiv2.Rule) model.Rule {
...
+	if ns != "" && (ar.Source.Selector != "" || ar.Source.NotSelector != "") {
+		logCxt := log.WithFields(log.Fields{
+			"Namespace": ns,
+			"Selector": ar.Source.Selector,
+			"NotSelector": ar.Source.NotSelector,
+		})
+		logCxt.Debug("Maybe update source Selector to include namespace")
+		if !strings.Contains(ar.Source.Selector, "kns.") {
+			logCxt.Debug("Updating source selector")
+			if ar.Source.Selector == "" {
```
Comment (caseydavenport) [inline lib/backend/syncersv1/updateprocessors/rules.go]: At this point we've already identified that `ar.Source.Selector != ""` so we can remove this check.

File: lib/backend/syncersv1/updateprocessors/rules.go [hunk 18]
```diff
@@ -46,6 +49,45 @@ func RuleAPIV2ToBackend(ar apiv2.Rule) model.Rule {
...
 		notICMPType = ar.NotICMP.Type
 	}
+	
+	// If we have any selector specified, then we may need to add the namespace selector.
+	// We do this if this policy is namespaced AND if the Selector does not have any other
+	// k8s namespace (profile label) selector in it.
+	// TODO this is TEMPORARY CODE:  We currently do a simple regex to search for kns. in the
+	// selector to see if we are performing k8s namespace queries.
+	nsSelector := fmt.Sprintf("%s == '%s'", apiv2.LabelNamespace, ns)
+	if ns != "" && (ar.Source.Selector != "" || ar.Source.NotSelector != "") {
```
Comment (caseydavenport) [inline lib/backend/syncersv1/updateprocessors/rules.go]: We're not doing anything with `NotSelector` below. I think we need to namespace that as well (if present) using the same logic as for `Selector` in order to be technically correct.

File: lib/backend/syncersv1/updateprocessors/rules.go (hunk 17 above)
Comment (robbrockbank) [reply lib/backend/syncersv1/updateprocessors/rules.go]: no we haven't

Comment (robbrockbank) [reply lib/backend/syncersv1/updateprocessors/rules.go]: first check is if either are != ""

And the ".kns" check is a negated check

Comment (caseydavenport) [reply lib/backend/syncersv1/updateprocessors/rules.go]: Yeah you're right. Carry on.

File: lib/backend/syncersv1/updateprocessors/rules.go (hunk 18 above)
Comment (caseydavenport) [reply lib/backend/syncersv1/updateprocessors/rules.go]: Discussed offline and this is fine.
